language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -328,11 +438,17 @@ inferTypeM (Constant _ (Some (ValueOf uni _))) =
     -- See Note [PLC types and universes].
     pure . Normalized . TyBuiltin () $ Some (TypeIn uni)
 
--- [infer| G !- bi : vTy]
--- ------------------------------
--- [infer| G !- builtin bi : vTy]
-inferTypeM (Builtin _ bi)           =
-    inferTypeOfBuiltinM bi
+
+-- bn : [a_1 :: K_1, ..., a_n:: K_m](B_1, ..., B_n)C
+-- G !- M_i : D_i
+-- D_i == [A_1,...,A_m/a_1,...,a_m]B_i
+----------------------------------------------------------------
+-- G !- ({builtin bn A_1, ..., A_m} : [A_1,...,A_m/a_1,...,a_m]C)
+inferTypeM (ApplyBuiltin ann bn tyargs args) = do
+  TypeComponents typeVars argTypes rtype <- inferTypeComponentsOfBuiltinM ann bn
+  rtype' <- checkBuiltinAppl ann bn typeVars tyargs argTypes args rtype
```

Comment:
I'd inline the definition of `checkBuiltinAppl` as it's just

1. check types
2. check terms
3. return the final type

which all seem to be very different things. Pulling the whole `do` into a separate function would also be sensible, then inlining `checkBuiltinAppl` would look even better. Right now `checkBuiltinAppl` receives way too many arguments loosely related to each other.